<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Resolve Bioscience Breast Cancer Subcellular â€¢ Giotto Suite</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Resolve Bioscience Breast Cancer Subcellular">
<meta property="og:description" content="Giotto">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BCXT5392QC"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BCXT5392QC');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto Suite</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting_started.html#installation">Installation</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#getting-started">Getting Started</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#analyses">Analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/datasets_overview.html#array-based-datasets-1">Array based</a>
    </li>
    <li>
      <a href="../articles/datasets_overview.html#single-cell-1">Single-cell</a>
    </li>
    <li>
      <a href="../articles/datasets_overview.html#subcellular-datasets-1">Subcellular</a>
    </li>
    <li>
      <a href="../articles/datasets_overview.html#protein-multiplexing-1">Protein multiplexing</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/contribute.html">Contribute</a>
</li>
<li>
  <a href="../articles/giotto_news.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    FAQ
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/faqs.html">Frequently Asked Questions</a>
    </li>
    <li>
      <a href="../articles/github_issues.html">GitHub Issues Guidelines</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Try Giotto
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Docker</li>
    <li class="dropdown-header">Binder</li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drieslab/Giotto/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Resolve Bioscience Breast Cancer
Subcellular</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/vignettes/Resolve_bc_210928.Rmd" class="external-link"><code>vignettes/Resolve_bc_210928.Rmd</code></a></small>
      <div class="hidden name"><code>Resolve_bc_210928.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="start-giotto">Start Giotto<a class="anchor" aria-label="anchor" href="#start-giotto"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. set working directory</span></span>
<span><span class="va">my_working_dir</span> <span class="op">=</span> <span class="st">'/path/to/directory'</span></span>
<span></span>
<span><span class="co"># 2. set giotto python path</span></span>
<span><span class="co"># set python path to your preferred python version path</span></span>
<span><span class="co"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span></span>
<span><span class="va">python_path</span> <span class="op">=</span> <span class="cn">NULL</span> </span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">python_path</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/installGiottoEnvironment.html">installGiottoEnvironment</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-files">Input Files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## provide path to resolve bioscience folder</span></span>
<span><span class="va">data_path</span> <span class="op">=</span> <span class="st">'/path/to/Resolve_bioscience_cancer'</span></span>
<span></span>
<span></span>
<span><span class="co"># 1. original image as png</span></span>
<span><span class="va">original_DAPI_image</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">'/'</span>, <span class="st">'sample2_DAPI.jpg'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. input cell segmentation as mask file</span></span>
<span><span class="co"># can also be provided as a 3-column polygon file</span></span>
<span><span class="co"># to be used as image background AND to store segmentations as polygons</span></span>
<span><span class="co"># can be obtained through Fiji / QuPath / Ilastik / Cellpose / ...</span></span>
<span><span class="va">segmentation_mask</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">'/'</span>, <span class="st">'Mask3.png'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. input features coordinates</span></span>
<span><span class="va">tx_coord</span> <span class="op">=</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data_path</span>, <span class="st">'/'</span>, <span class="st">'data_sample2.txt'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/row-plus-colnames-generic.html">colnames</a></span><span class="op">(</span><span class="va">tx_coord</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x'</span>, <span class="st">'y'</span>, <span class="st">'z'</span>, <span class="st">'gene_id'</span><span class="op">)</span></span>
<span><span class="va">tx_coord</span> <span class="op">=</span> <span class="va">tx_coord</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">gene_id</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-1-create-subcellular-giotto-object">Part 1: Create Subcellular Giotto Object<a class="anchor" aria-label="anchor" href="#part-1-create-subcellular-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoObjectSubcellular.html">createGiottoObjectSubcellular</a></span><span class="op">(</span>gpoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="va">tx_coord</span><span class="op">)</span>,</span>
<span>                                        gpolygons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'cell'</span> <span class="op">=</span> <span class="va">segmentation_mask</span><span class="op">)</span>,</span>
<span>                                        instructions <span class="op">=</span> <span class="va">instrs</span>,</span>
<span>                                        verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                        cores <span class="op">=</span> <span class="fl">32</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-2-create-spatial-locations">Part 2: Create Spatial Locations<a class="anchor" aria-label="anchor" href="#part-2-create-spatial-locations"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># centroids are now used to provide the spatial locations (centroid of each cell)</span></span>
<span><span class="co"># needed for certain downstream spatial analyses</span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/addSpatialCentroidLocations.html">addSpatialCentroidLocations</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                                      poly_info <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-3-add-image-information">Part 3: Add Image Information<a class="anchor" aria-label="anchor" href="#part-3-add-image-information"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create Giotto images</span></span>
<span><span class="va">DAPI_image</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoImage.html">createGiottoImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">'DAPI'</span>,</span>
<span>                               do_manual_adj <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                               xmax_adj <span class="op">=</span> <span class="fl">0</span>,ymax_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               xmin_adj <span class="op">=</span> <span class="fl">0</span>,ymin_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               image_transformations <span class="op">=</span> <span class="st">'flip_x_axis'</span>,</span>
<span>                               mg_object <span class="op">=</span> <span class="va">original_DAPI_image</span><span class="op">)</span></span>
<span></span>
<span><span class="va">segm_image</span> <span class="op">=</span> <span class="fu"><a href="../reference/createGiottoImage.html">createGiottoImage</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>,</span>
<span>                               name <span class="op">=</span> <span class="st">'segmentation'</span>,</span>
<span>                               do_manual_adj <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                               xmax_adj <span class="op">=</span> <span class="fl">0</span>,ymax_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               xmin_adj <span class="op">=</span> <span class="fl">0</span>,ymin_adj <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                               image_transformations <span class="op">=</span> <span class="st">'flip_x_axis'</span>,</span>
<span>                               mg_object <span class="op">=</span> <span class="va">segmentation_mask</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add images to Giotto object</span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/addGiottoImage.html">addGiottoImage</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                         images <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">DAPI_image</span>, <span class="va">segm_image</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># provides an overview of available images</span></span>
<span><span class="fu"><a href="../reference/showGiottoImageNames.html">showGiottoImageNames</a></span><span class="op">(</span><span class="va">testobj</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-4-visualize-original-images">Part 4: Visualize Original Images<a class="anchor" aria-label="anchor" href="#part-4-visualize-original-images"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize overlay of calculated cell centroid with original image and segmentation mask file</span></span>
<span><span class="co"># by setting show_plot to FALSE and save_plot to TRUE you can save quite some time when creating plots</span></span>
<span><span class="co"># with big images it sometimes takes quite long for R/Rstudio to render them</span></span>
<span><span class="fu"><a href="../reference/spatPlot2D.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>, image_name <span class="op">=</span> <span class="st">'DAPI'</span>, point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/0-spatPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatPlot2D.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>, image_name <span class="op">=</span> <span class="st">'segmentation'</span>, point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/1-spatPlot2D.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="part-5-calculate-cell-shape-overlap">Part 5: Calculate Cell Shape Overlap<a class="anchor" aria-label="anchor" href="#part-5-calculate-cell-shape-overlap"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu">calculateOverlap</span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                           method <span class="op">=</span> <span class="st">'parallel'</span>,</span>
<span>                           x_step <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                           y_step <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                           poly_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                           feat_info <span class="op">=</span> <span class="st">'rna'</span><span class="op">)</span></span>
<span><span class="fu">tictoc</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#convert overlap to matrix </span></span>
<span><span class="va">testobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/overlapToMatrix.html">overlapToMatrix</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                          poly_info <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                          feat_info <span class="op">=</span> <span class="st">'rna'</span>,</span>
<span>                          name <span class="op">=</span> <span class="st">'raw'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-6-filter-data">Part 6: Filter Data<a class="anchor" aria-label="anchor" href="#part-6-filter-data"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># features can be filtered individually</span></span>
<span><span class="co"># cells will be filtered across features</span></span>
<span></span>
<span><span class="co"># first filter on rna</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterGiotto.html">filterGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">testobj</span>,</span>
<span>                          expression_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          feat_det_in_min_cells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>                          min_det_feats_per_cell <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/spatPlot2D.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>           image_name <span class="op">=</span> <span class="st">'segmentation'</span>, show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/69-spatPlot2D.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="part-7-process-giotto-object">Part 7: Process Giotto Object<a class="anchor" aria-label="anchor" href="#part-7-process-giotto-object"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rna data, default.</span></span>
<span><span class="co"># other feature modalities can be processed and filtered in an anologous manner</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, scalefactor <span class="op">=</span> <span class="fl">6000</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatistics.html">addStatistics</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span><span class="op">)</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjustGiottoMatrix.html">adjustGiottoMatrix</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                                expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'normalized'</span><span class="op">)</span>,</span>
<span>                                covariate_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'nr_feats'</span>, <span class="st">'total_expr'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalizeGiotto.html">normalizeGiotto</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, norm_methods <span class="op">=</span> <span class="st">'pearson_resid'</span>, update_slot <span class="op">=</span> <span class="st">'pearson'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/showGiottoExpression.html">showGiottoExpression</a></span><span class="op">(</span><span class="va">subc_test</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="part-8-dimension-reduction">Part 8: Dimension Reduction<a class="anchor" aria-label="anchor" href="#part-8-dimension-reduction"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Find highly valuable Features</span></span>
<span></span>
<span><span class="co"># typical way of calculating HVF</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, HVFname<span class="op">=</span> <span class="st">'hvg_orig'</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/4-HVFplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># new method based on variance of pearson residuals for each gene</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateHVF.html">calculateHVF</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                          method <span class="op">=</span> <span class="st">'var_p_resid'</span>, expression_values <span class="op">=</span> <span class="st">'pearson'</span>,</span>
<span>                          show_plot <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/5-HVFplot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#run PCA</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runPCA.html">runPCA</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                    expression_values <span class="op">=</span> <span class="st">'pearson'</span>,</span>
<span>                    scale_unit <span class="op">=</span> <span class="cn">F</span>, center <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/screePlot.html">screePlot</a></span><span class="op">(</span><span class="va">subc_test</span>, ncp <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/6-screePlot.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotPCA.html">plotPCA</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>        dim1_to_use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        dim2_to_use <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/7-PCA.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run UMAP</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runUMAP.html">runUMAP</a></span><span class="op">(</span><span class="va">subc_test</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, n_threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/8-UMAP.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="part-9-cluster">Part 9: Cluster<a class="anchor" aria-label="anchor" href="#part-9-cluster"></a>
</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createNearestNetwork.html">createNearestNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, dimensions_to_use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">subc_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doLeidenCluster.html">doLeidenCluster</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, resolution <span class="op">=</span> <span class="fl">0.05</span>, n_iterations <span class="op">=</span> <span class="fl">1000</span>, name <span class="op">=</span> <span class="st">'leiden_0.05'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create color palettes, or proceed with Giotto defaults</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"alyssafrazee/RSkittleBrewer"</span><span class="op">)</span></span>
<span><span class="va">colorcode</span> <span class="op">=</span> <span class="fu">lacroix_palette</span><span class="op">(</span>type <span class="op">=</span> <span class="st">"paired"</span><span class="op">)</span></span>
<span><span class="va">featcolor</span> <span class="op">=</span> <span class="fu">lacroix_palette</span><span class="op">(</span><span class="st">"KeyLime"</span>, type <span class="op">=</span> <span class="st">"discrete"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># visualize UMAP cluster results</span></span>
<span><span class="fu"><a href="../reference/plotUMAP.html">plotUMAP</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>         show_NN_network <span class="op">=</span> <span class="cn">T</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>, cell_color_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/39-UMAP.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># visualize UMAP and spatial results</span></span>
<span><span class="fu"><a href="../reference/spatDimPlot2D.html">spatDimPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>              show_image <span class="op">=</span> <span class="cn">T</span>, image_name <span class="op">=</span> <span class="st">'segmentation'</span>,</span>
<span>              cell_color <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>              spat_point_size <span class="op">=</span> <span class="fl">2</span>, cell_color_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/51-spatDimPlot2D.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot a cluster heatmap</span></span>
<span><span class="fu"><a href="../reference/showClusterHeatmap.html">showClusterHeatmap</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>, cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                   save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>save_format <span class="op">=</span> <span class="st">'pdf'</span>,base_height <span class="op">=</span> <span class="fl">6</span>, base_width <span class="op">=</span> <span class="fl">8</span>, units <span class="op">=</span> <span class="st">'cm'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/72-Heatmap.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># See cluster relationships in a dendogram</span></span>
<span><span class="fu"><a href="../reference/showClusterDendrogram.html">showClusterDendrogram</a></span><span class="op">(</span><span class="va">subc_test</span>, h <span class="op">=</span> <span class="fl">0.5</span>, rotate <span class="op">=</span> <span class="cn">T</span>, cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/12-showClusterDendrogram.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="part-10-create-a-spatial-network">Part 10: Create a Spatial Network<a class="anchor" aria-label="anchor" href="#part-10-create-a-spatial-network"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subc_test</span> <span class="op">=</span> <span class="fu"><a href="../reference/createSpatialNetwork.html">createSpatialNetwork</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                                 spat_loc_name <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                                 minimum_k <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                 maximum_distance_delaunay <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/spatPlot2D.html">spatPlot2D</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>           image_name <span class="op">=</span> <span class="st">'segmentation'</span>, show_image <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           point_size <span class="op">=</span> <span class="fl">1.5</span>, show_network <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/13-spatPlot2D.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="part-11-visualize-subcellular-data">Part 11: Visualize SubCellular Data<a class="anchor" aria-label="anchor" href="#part-11-visualize-subcellular-data"></a>
</h2>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize clustered cells</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span> , </span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/54-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># individual plotting of transcripts and polygon information</span></span>
<span></span>
<span><span class="co"># all cells</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">testobj</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/16-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># filtered cells</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/17-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># faster plotting method if you have many points</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     plot_method <span class="op">=</span> <span class="st">'scattermore'</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/18-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="section level3">
<h3 id="subset-by-loaction">Subset by Loaction<a class="anchor" aria-label="anchor" href="#subset-by-loaction"></a>
</h3>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># can be used to focus on specific spatial structures</span></span>
<span><span class="co"># to zoom in on niche environments</span></span>
<span></span>
<span><span class="va">subloc</span> <span class="op">=</span> <span class="fu"><a href="../reference/subsetGiottoLocs.html">subsetGiottoLocs</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                          x_min <span class="op">=</span> <span class="fl">0</span>, x_max <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>                          y_min <span class="op">=</span> <span class="fl">0</span>, y_max <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>                          poly_info <span class="op">=</span> <span class="st">'cell'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show subset of genes</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/19-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show subset of genes and color cells according to clusters</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span>,</span>
<span>                     feats_color_code <span class="op">=</span> <span class="va">featcolor</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/61-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show subset of genes and color cells according to total expression</span></span>
<span><span class="co"># use a faster and more efficient point plotting method = scattermore</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     plot_method <span class="op">=</span> <span class="st">'scattermore'</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MMP2"</span>, <span class="st">"VEGFA"</span>, <span class="st">"IGF1R"</span>, <span class="st">'CDH2'</span>, <span class="st">'MKI67'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'total_expr'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/21-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show cells and color them according to total expression</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'total_expr'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/22-spatInSituPlotPoints.png" style="width:50.0%"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show cells and color them according to total cluster information</span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subloc</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'white'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/66-spatInSituPlotPoints.png" style="width:50.0%"></p>
</div>
</div>
<div class="section level2">
<h2 id="part-12-find-interaction-changed-genes">Part 12: Find Interaction Changed Genes<a class="anchor" aria-label="anchor" href="#part-12-find-interaction-changed-genes"></a>
</h2>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find interaction changed genes</span></span>
<span><span class="co"># genes whose expression difference is associated with a neighboring cell type</span></span>
<span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">'multisession'</span>, workers <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="co"># sometimes unstable, restart R session</span></span>
<span></span>
<span><span class="va">test</span> <span class="op">=</span> <span class="fu"><a href="../reference/findInteractionChangedFeats.html">findInteractionChangedFeats</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">subc_test</span>,</span>
<span>                                   cluster_column <span class="op">=</span> <span class="st">'leiden_clus'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">test</span><span class="op">$</span><span class="va">CPGscores</span><span class="op">[</span><span class="va">type_int</span> <span class="op">==</span> <span class="st">'hetero'</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="../reference/spatInSituPlotPoints.html">spatInSituPlotPoints</a></span><span class="op">(</span><span class="va">subc_test</span>,</span>
<span>                     feats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'rna'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTSD"</span>, <span class="st">"BMP1"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                     point_size <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>                     show_polygon <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     polygon_feat_type <span class="op">=</span> <span class="st">'cell'</span>,</span>
<span>                     polygon_color <span class="op">=</span> <span class="st">'black'</span>,</span>
<span>                     polygon_line_size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                     polygon_fill <span class="op">=</span> <span class="st">'leiden_clus'</span>,</span>
<span>                     polygon_fill_as_factor <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                     polygon_fill_code <span class="op">=</span> <span class="va">colorcode</span>,</span>
<span>                     feats_color_code <span class="op">=</span> <span class="va">featcolor</span><span class="op">)</span></span></code></pre></div>
<p><img src="../inst/images/Resolve_bc/210928/43-spatInSituPlotPoints.png" style="width:50.0%"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.drieslab.com/" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, Emma Kelley, Junxiang Xu, Guo-Cheng Yuan, Matthew Oâ€™Brien, Joselyn ChÃ¡vez.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
