<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>plotMetaDataHeatmap — plotMetaDataHeatmap • Giotto Suite</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="plotMetaDataHeatmap — plotMetaDataHeatmap"><meta property="og:description" content="Creates heatmap for genes within aggregated clusters."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-BCXT5392QC"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-BCXT5392QC');
</script></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Giotto Suite</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Start
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/getting_started.html#installation">Installation</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#getting-started">Getting Started</a>
    </li>
    <li>
      <a href="../articles/getting_started.html#analyses">Analyses</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Datasets
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/datasets_overview.html#array-based-datasets-1">Array based</a>
    </li>
    <li>
      <a href="../articles/datasets_overview.html#single-cell-1">Single-cell</a>
    </li>
    <li>
      <a href="../articles/datasets_overview.html#subcellular-datasets-1">Subcellular</a>
    </li>
    <li>
      <a href="../articles/datasets_overview.html#protein-multiplexing-1">Protein multiplexing</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/contribute.html">Contribute</a>
</li>
<li>
  <a href="../articles/giotto_news.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    FAQ
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/faqs.html">Frequently Asked Questions</a>
    </li>
    <li>
      <a href="../articles/github_issues.html">GitHub Issues Guidelines</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Try Giotto
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Docker</li>
    <li class="dropdown-header">Binder</li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drieslab/Giotto/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>plotMetaDataHeatmap</h1>
    <small class="dont-index">Source: <a href="https://github.com/drieslab/Giotto/blob/HEAD/R/auxiliary_visuals.R" class="external-link"><code>R/auxiliary_visuals.R</code></a></small>
    <div class="hidden name"><code>plotMetaDataHeatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Creates heatmap for genes within aggregated clusters.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotMetaDataHeatmap</span><span class="op">(</span></span>
<span>  <span class="va">gobject</span>,</span>
<span>  spat_unit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feat_type <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expression_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"normalized"</span>, <span class="st">"scaled"</span>, <span class="st">"custom"</span><span class="op">)</span>,</span>
<span>  metadata_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  selected_feats <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  selected_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  first_meta_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  second_meta_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  show_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"zscores"</span>, <span class="st">"original"</span>, <span class="st">"zscores_rescaled"</span><span class="op">)</span>,</span>
<span>  custom_cluster_order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clus_cor_method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  clus_cluster_method <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>  custom_feat_order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_gene_order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feat_cor_method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  gene_cor_method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  feat_cluster_method <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>  gene_cluster_method <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gradient_color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>  gradient_midpoint <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  gradient_limits <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  x_text_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  x_text_angle <span class="op">=</span> <span class="fl">45</span>,</span>
<span>  y_text_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  strip_text_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  show_plot <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  return_plot <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  save_plot <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  save_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  default_save_name <span class="op">=</span> <span class="st">"plotMetaDataHeatmap"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gobject</dt>
<dd><p>giotto object</p></dd>


<dt>spat_unit</dt>
<dd><p>spatial unit (e.g. "cell")</p></dd>


<dt>feat_type</dt>
<dd><p>feature type (e.g. "rna", "dna", "protein")</p></dd>


<dt>expression_values</dt>
<dd><p>expression values to use (e.g. "normalized", "scaled", "custom")</p></dd>


<dt>metadata_cols</dt>
<dd><p>annotation columns found in pDataDT(gobject)</p></dd>


<dt>selected_feats</dt>
<dd><p>subset of features to use</p></dd>


<dt>selected_genes</dt>
<dd><p>deprecated. See <code>selected_feats</code> param</p></dd>


<dt>first_meta_col</dt>
<dd><p>if more than 1 metadata column, select the x-axis factor</p></dd>


<dt>second_meta_col</dt>
<dd><p>if more than 1 metadata column, select the facetting factor</p></dd>


<dt>show_values</dt>
<dd><p>which values to show on heatmap (e.g. "zscores", "original", "zscores_rescaled")</p></dd>


<dt>custom_cluster_order</dt>
<dd><p>custom cluster order (default = NULL)</p></dd>


<dt>clus_cor_method</dt>
<dd><p>correlation method for clusters, default to "pearson"</p></dd>


<dt>clus_cluster_method</dt>
<dd><p>hierarchical cluster method for the clusters, default to "complete"</p></dd>


<dt>custom_feat_order</dt>
<dd><p>custom feature order (default = NULL)</p></dd>


<dt>custom_gene_order</dt>
<dd><p>deprecated. See <code>custom_feat_order</code> param</p></dd>


<dt>feat_cor_method</dt>
<dd><p>correlation method for features, default to "pearson"</p></dd>


<dt>gene_cor_method</dt>
<dd><p>deprecated. See <code>feat_cor_method</code> param</p></dd>


<dt>feat_cluster_method</dt>
<dd><p>hierarchical cluster method for the features, default to "complete"</p></dd>


<dt>gene_cluster_method</dt>
<dd><p>deprecated. See <code>feat_cluster_method</code> param</p></dd>


<dt>gradient_color</dt>
<dd><p>vector with 3 colors for numeric data</p></dd>


<dt>gradient_midpoint</dt>
<dd><p>midpoint for color gradient</p></dd>


<dt>gradient_limits</dt>
<dd><p>vector with lower and upper limits</p></dd>


<dt>x_text_size</dt>
<dd><p>size of x-axis text</p></dd>


<dt>x_text_angle</dt>
<dd><p>angle of x-axis text</p></dd>


<dt>y_text_size</dt>
<dd><p>size of y-axis text</p></dd>


<dt>strip_text_size</dt>
<dd><p>size of strip text</p></dd>


<dt>show_plot</dt>
<dd><p>show plot. TRUE or FALSE</p></dd>


<dt>return_plot</dt>
<dd><p>return ggplot object. TRUE or FALSE</p></dd>


<dt>save_plot</dt>
<dd><p>directly save the plot. TRUE or FALSE</p></dd>


<dt>save_param</dt>
<dd><p>list of saving parameters, see <code><a href="showSaveParameters.html">showSaveParameters</a></code></p></dd>


<dt>default_save_name</dt>
<dd><p>default save name</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>ggplot or data.table</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Creates heatmap for the average expression of selected genes in the different annotation/cluster groups.
Calculation of cluster or gene order is done on the provided expression values, but visualization
is by default on the z-scores. Other options are the original values or z-scores rescaled per gene (-1 to 1).</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="plotMetaDataCellsHeatmap.html">plotMetaDataCellsHeatmap</a></code> for numeric cell annotation instead of gene expression.</p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.drieslab.com/" class="external-link">Ruben Dries</a>, Jiaji George Chen, Natalie Del Rossi, Emma Kelley, Junxiang Xu, Guo-Cheng Yuan, Matthew O’Brien, Joselyn Chávez.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

